### ISSUE TITLE
Include annotations along with method names to allow filtering

### ISSUE THREAD

**Issue by kirillzh** [linkedin/dex-test-parser#1]
Since instrumentation allows filtering by annotation, it would be useful to include list of all annotations for each test method and the whole test class.

Also, in what format should the data be stored? CSV with different separators for columns and lists, or maybe JSON?

**Comment by sander-m** [comment]
Any update on this? This would indeed be useful.

**Comment by kirillzh** [comment]
@sander-m, I've implemented it on my fork (https://github.com/kirillzh/dex-test-parser/commit/9b8d473d24a9f91dff099be835e2680733d015b8). I had to disable JUnit 3 tests though (I don't need them). Tests will be saved JSON format:
```json
[
   {
      "class":"com.app.tests.SomeFeatureTests",
      "method":"fooTest",
      "annotations":[
         "com.app.annotations.Smoke",
         "org.junit.Ignore",
         "org.junit.Test"
      ]
   },
   {
      "class":"com.app.tests.SomeOtherFeatureTests",
      "method":"barTest",
      "annotations":[
         "com.app.annotations.Regression",
         "org.junit.Test"
      ]
   }
]
```

I'd create a PR but I didn't get any response in this ticket so I'm not sure what requirements are for the output.

**Comment by drewhannay** [comment]
Sorry for not responding earlier....I'm not quite clear what the benefit of this would be. We do lots of filtering internally, but we just run the test anyway and let AndroidJUnitRunner decide to run or skip the test. The benefit of this approach is that the skipped tests will still show up in the JUnit test report as skipped, rather than missing entirely.

**Comment by kirillzh** [comment]
No problem!

So how do you guys group your test cases? For example if I categorize my test cases as `smoke`, `regression`, etc. or maybe by `feature A`, `feature B`, etc., how do I run only `smoke` tests, or only `feature A` tests?

**Comment by drewhannay** [comment]
We group our tests by package and run them that way (e.g. `com.example.smoke.featureA`, `com.example.regression.featureB`. Then if you wanted to run all the featureA tests, you'd just regex for `.featureA.`, or similarly package=`com.example.smoke` and run all the matching tests individually.

I was thinking you could choose to run a specific test (e.g. `-e class com.example.smoke.MyTest#foo`) and _also_ specify an annotation filter (e.g. `-e annotation com.android.foo.MyAnnotation`)

Now that I think about it, I'm not sure if that would actually work or not. I'd be hopeful that it does, and it won't trigger a classpath scan by `AndroidJUnitRunner`, but if that doesn't work, then I understand why this feature would be needed.

**Comment by kirillzh** [comment]
I see, `com.example.smoke.featureA` and `com.example.regression.featureB` is a pretty good idea, didn't think about it haha.
I group my tests by feature only though, so it's easier for me to convert test say from smoke to regression simply by changing annotation instead of moving it from one class to another.

And you're right, I could just run all tests with annotation filter, it won't run the tests without this annotation:
```
INSTRUMENTATION_RESULT: stream=

Time: 0

OK (0 tests)

INSTRUMENTATION_CODE: -1
```
But I guess I'd have to parse and ignore this test in reports somehow...

**Comment by sander-m** [comment]
Thanks @kirillzh I will have a look at that branch.

Currently our tests use class-level annotations to group them by type (Unit/Integration/UI etc), and I plan to use this dex-parsing for sharding test runs with Firebase Test Lab(https://cloud.google.com/sdk/gcloud/reference/firebase/test/android/run), which doesn't take annotation filters.

**Comment by bootstraponline** [comment]
It would be really nice to have annotations support. As mentioned, the Firebase Test Lab expects a list of tests to run.

**Comment by drewhannay** [comment]
I'm not opposed to adding that if it's useful. Do you all think it's a requirement to have it write the data to a file? If we only expose this through the Java/Kotlin api, then we don't need to worry about the file format.

I'm guessing you'll want it as a file though, if your test harness isn't written in a JVM language (for us, our test harness is a gradle plugin, so we don't even use the file output from this project).

**Comment by bootstraponline** [comment]
I just checked with the Firebase team and they support filtering by annotations. It's just undocumented.

> adb shell am instrument -w -e annotation com.android.foo.MyAnnotation

gets translated to something like:

> --environment-variables annotation=com.android.foo.MyAnnotation

- [AndroidJUnitRunner](https://developer.android.com/reference/android/support/test/runner/AndroidJUnitRunner.html)
- [running subset of espresso tests](https://medium.com/@satyajit/running-subset-of-espresso-tests-locally-and-gcloud-4713931da56)

--- 

> If we only expose this through the Java/Kotlin API

I think exposing via the API makes sense. It's not that difficult to create a wrapper jar which outputs in the desired format.

**Comment by bootstraponline** [comment]
I found a use case. The default [AndroidJUnitRunner](https://github.com/bootstraponline/platform_frameworks_testing/blob/android-support-test/runner/src/main/java/android/support/test/internal/runner/TestRequestBuilder.java#L718) only allows running a single annotation. If you want to run multiple annotations (SmallTest & LargeTest), then you'd have to write a custom test runner.

If the `dex-test-parser` API offered annotation support, then it'd be easy to build a list of tests using custom logic to mix and match instead of being limited by the test runner.

**Comment by drewhannay** [comment]
I believe I have this working locally for both JUnit 3 and JUnit 4. I'll put up a PR either tomorrow or sometime next week, after doing some more testing. I might implement #5 first, to have better regression coverage, because this is a fairly significant change.

One thing to note, gathering all this extra annotation information _does_ have a performance overhead. I've been testing against an apk with ~5000 JUnit 4 test methods (I don't currently have a large scale JUnit 3 test apk to profile against). Prior to these annotation changes, dex-test-parser took ~300ms to process this apk. With annotation collection, it takes ~500ms. I'm assuming that 5k methods is on the high end of test apk sizes, so 500ms is not that bad of an absolute time, but I don't have many data points for how these times scale for others (especially JUnit 3 users).

I think this is probably fine, given the benefits of this feature, but just wanted to make sure it's called out here since I know I monitor my own build and test speeds very closely.

**Comment by drewhannay** [comment]
This is checked in and released in 1.1.0! 😄

**Comment by bootstraponline** [comment]
Awesome. Thank you!!
